NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/mcx.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "mcx" 1 "16 May 2014" "mcx 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
mcx \- perform various computations on graphs and matrices
.SH SYNOPSIS

.di ZV
.in 0
.nf \fC
   \fBmcx\fP <mode> [mode-options] [mode-files]
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

.SH DESCRIPTION

\fBmcx\fP implements a variety of computations on graphs and matrices\&. The
first argument to \fBmcx\fP should be a \fImode\fP, which is a string establishing
the type of computation to invoke\&. Each mode \fImode\fP is described in the
mcx\fImode\fP manual page\&. The currently available modes are
\fBconvert\fP, \fBdiameter\fP, \fBclcf\fP, \fBq\fP and \fBctty\fP\&.
The \fBconvert\fP mode is thus described in the
\fBmcxconvert\fP manual page\&.

Invoking \fBmcx\fP without arguments causes it to print out a list
of available modes with a short description of the type of
command line expected by that mode\&.

Invoking \fBmcx\fP with just a mode will print out a longer listing
of options available for that particular mode\&.
Modes that are able to function normally without arguments
can be invoked by adding the \fI--nop\fP argument\&.

Several options are shared between all \fBmcx\fP modes\&. Note that
these options are specified after the \fBmode\fP\&. See \fBOPTIONS\fP\&.

The \fB--version\fP causes \fBmcx\fP to print out version
and license information\&.

Issuing \fImcx help mode\fP will cause \fBmcx\fP to look for a manual page
describing \fImode\fP and display it if found\&. This requires that the
MANPATH environment variable contains the directory in which the MCL-edge
manual pages were installed\&. This will usually be \fCPREFIX/share\fP,
where \fCPREFIX\fP is the path with which the software was configured\&.
.SH OPTIONS

These are options that pertain to all modes\&. They should be specified
\fIafter\fP the mode argument\&.

.ZI 2m "\fB-h\fP (\fIsynopsis\fP)"
\&
'in -2m
.ZI 2m "\fB--help\fP (\fIsynopsis\fP)"
\&
'in -2m
'in +2m
\&
.br
List available options\&.
.in -2m

.ZI 2m "\fB--nop\fP (\fIno-op\fP)"
\&
.br
Not an option\&. This option has no effect then to increment
the argument count\&. This can be useful for \fBmcx\fP modes which are able to
function without any options\&. Such a mode typically reads from STDIN, writes
to STDOUT, and uses default settings\&. However, simply specifying a mode
without options leads \fBmcx\fP to output a list of available options for that
mode\&. This can be prevented by using the \fB--nop\fP option\&.
.in -2m

.ZI 2m "\fB-set\fP (\fIkey=val\fP)"
\&
.br
Sets the key\ \&\fBkey\fP to value\ \&\fBval\fP in the environment\&.
Some modes allow adjustment of settings in this manner\&.
.in -2m

.ZI 2m "\fB-progress\fP <num> (\fIprogress interval size\fP)"
\&
.br
Defines the interval that defines the progress frequency\&.
.in -2m

.ZI 2m "\fB--debug\fP (\fIturn on debugging\fP)"
\&
.br
Turn debugging on\&.
.in -2m

.ZI 2m "\fB--version\fP (\fIshow version\fP)"
\&
.br
Show version\&.
.in -2m
.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shmx/mcx.c
/*   (C) Copyright 1999, 2000, 2001, 2002, 2003, 2004 Stijn van Dongen
 *   (C) Copyright 2005, 2006, 2007, 2008, 2009, 2010, 2011 Stijn van Dongen
 *   (C) Copyright 2012, 2013 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/


#include <string.h>
#include <stdio.h>

#include "mcx.h"

#include "mcxconvert.h"
#include "mcxquery.h"
#include "mcxdiameter.h"
#include "mcxerdos.h"
#include "mcxclcf.h"
#include "mcxcollect.h"
#include "mcxtab.h"
#include "mcxfp.h"
#include "mcxalter.h"

#include "impala/stream.h"
#include "impala/matrix.h"
#include "impala/tab.h"
#include "impala/io.h"
#include "impala/iface.h"
#include "impala/compose.h"
#include "impala/ivp.h"
#include "impala/app.h"

#include "util/types.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/ding.h"


const char* mcx_syntax =
   "Usage: mcx <mode> [options] [files]\n"
   "       mcx <mode> -h  (for mode specific options)"
   ;


enum
{  ID_CONVERT = 0
,  ID_QUERY
,  ID_ERDOS
,  ID_CTTY
,  ID_DIAMETER
,  ID_CLCF
,  ID_COLLECT
,  ID_TAB
,  ID_ALTER
,  ID_FP
,  ID_UNUSED
}  ;


mcxDispEntry mcx_dir[] =
{  {  ID_CONVERT  ,  mcxDispHookConvert      }
,  {  ID_QUERY    ,  mcxDispHookquery        }
,  {  ID_DIAMETER ,  mcxDispHookDiameter     }
,  {  ID_CLCF     ,  mcxDispHookClcf         }
,  {  ID_ERDOS    ,  mcxDispHookErdos        }
,  {  ID_CTTY     ,  mcxDispHookCtty         }
,  {  ID_COLLECT  ,  mcxDispHookCollect      }
,  {  ID_ALTER    ,  mcxDispHookAlter        }
,  {  ID_TAB      ,  mcxDispHookTab          }
,  {  ID_FP       ,  mcxDispHookFp           }
,  {  -1          ,  NULL                    }
}  ;


mcxOptAnchor mcxSharedOptions[] =
{  {  "--version"
   ,  MCX_OPT_DEFAULT
   ,  MCX_DISP_VERSION
   ,  NULL
   ,  "output version information, exit"
   }
,  {  "--test"
   ,  MCX_OPT_DEFAULT | MCX_OPT_HIDDEN
   ,  MCX_DISP_TEST
   ,  NULL
   ,  "test"
   }
,  {  "-progress"
   ,  MCX_OPT_HASARG
   ,  MCX_DISP_PROGRESS
   ,  "<int>"
   ,  "set progress interval size"
   }
,  {  "--progress"
   ,  MCX_OPT_DEFAULT
   ,  MCX_DISP_PROGRESS2
   ,  NULL
   ,  "turn on progress reporting"
   }
,  {  "-debug"
   ,  MCX_OPT_HASARG | MCX_OPT_HIDDEN
   ,  MCX_DISP_DEBUG
   ,  "<int>"
   ,  "set debug level or bits"
   }
,  {  "--debug"
   ,  MCX_OPT_DEFAULT | MCX_OPT_HIDDEN
   ,  MCX_DISP_DEBUG2
   ,  NULL
   ,  "turn on debugging"
   }
,  {  "-set"
   ,  MCX_OPT_HASARG
   ,  MCX_DISP_SET
   ,  "key=val"
   ,  "set key to val in ENV"
   }
,  {  "--nop"
   ,  MCX_OPT_DEFAULT
   ,  MCX_DISP_NOP
   ,  NULL
   ,  "this option has no affect but changing the argument count"
   }
,  {  "--write-binary"
   ,  MCX_OPT_DEFAULT
   ,  MCX_DISP_WB
   ,  NULL
   ,  "write output in binary format"
   }
,  {  "-h"
   ,  MCX_OPT_DEFAULT
   ,  MCX_DISP_HELP
   ,  NULL
   ,  "this"
   }
,  {  "--help"
   ,  MCX_OPT_DEFAULT
   ,  MCX_DISP_APROPOS
   ,  NULL
   ,  "this"
   }
,  {  "--amoixa"
   ,  MCX_OPT_HIDDEN
   ,  MCX_DISP_AMOIXA
   ,  NULL
   ,  "show hidden options too"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


mcxbits  mcx_debug_g     =  0;
unsigned mcx_progress_g  =  0;
mcxbool  mcx_test_g      =  FALSE;
mcxbool  mcx_wb_g =  FALSE;


mcxstatus sharedArgHandle
(  int optid
,  const char* val
,  mcxDispHook*   hook
,  mcxDispBundle* bundle
)
   {  mcxTing* full_syntax = mcxTingPrint(NULL, "%s %s", "mcx", hook->syntax)
   ;  switch(optid)
      {  case MCX_DISP_HELP
      :  case MCX_DISP_APROPOS
      :  case MCX_DISP_AMOIXA
      :  mcxOptApropos
         (  stdout
         ,  hook->name
         ,  full_syntax->str
         ,  15
         ,  0
         ,  bundle->disp_shared
         )
      ;  mcxOptApropos
         (  stdout
         ,  hook->name
         ,  NULL
         ,  15
         ,     MCX_OPT_DISPLAY_SKIP
            |  (  optid == MCX_DISP_AMOIXA
               ?  MCX_OPT_DISPLAY_HIDDEN
               :  0
               )
         ,  hook->options
         )
      ;  mcxExit(0)
      ;  break
      ;

         case MCX_DISP_VERSION
      :  bundle->disp_version(hook->name)
      ;  mcxExit(0)
      ;  break
      ;

         case MCX_DISP_NOP
      :  NOTHING
      ;  break
      ;

         case MCX_DISP_WB
      :  mcx_wb_g = TRUE
      ;  break
      ;

         case MCX_DISP_TEST
      :  mcx_test_g = TRUE
      ;  break
      ;

         case MCX_DISP_PROGRESS
      :  mcx_progress_g = atoi(val)
      ;  break
      ;

         case MCX_DISP_PROGRESS2
      :  mcx_progress_g = 1
      ;  break
      ;

         case MCX_DISP_DEBUG
      :  mcx_debug_g = atoi(val)
      ;  break
      ;

         case MCX_DISP_DEBUG2
      :  mcx_debug_g = -1u
      ;  break
      ;

         case MCX_DISP_SET
      :  mcxSetenv(val)
      ;  break
      ;

         default
      :  break
      ;
      }
      mcxTingFree(&full_syntax)
   ;  return STATUS_OK
;  }


int main
(  int                  argc
,  const char*          argv[]
)
   {  mcxDispBundle bundle
   ;  mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_YES)

   ;  mcxLogLevel =
      MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_IO | MCX_LOG_GAUGE | MCX_LOG_WARN

   ;  mclx_app_init(stderr)

   ;  bundle.disp_argc     =  argc
   ;  bundle.disp_argv     =  argv
   ;  bundle.disp_name     =  "mcx"
   ;  bundle.disp_syntax   =  mcx_syntax
   ;  bundle.disp_shared   =  mcxSharedOptions
   ;  bundle.n_disp_shared =  sizeof(mcxSharedOptions) / sizeof(mcxOptAnchor) -1;
   ;  bundle.disp_table    =  mcx_dir
   ;  bundle.disp_version  =  app_report_version
   ;  bundle.shared_handler=  sharedArgHandle
      
   ;  return mcxDispatch(&bundle)
;  }


mclx* mcx_get_graph
(  const char* caller
,  mcxIO*   xfmx
,  mcxIO*   xfabc
,  mcxIO*   xftab
,  mclTab** tabpp
,  mclgTF*  transform
,  mcxbits  readx_bits
)
   {  mclTab* tab = tabpp[0]
   ;  mclx* mx = NULL
   ;  mclxIOstreamer streamer = { 0 }

   ;  if (xfabc)
      {  if (xftab)
         {  tab = mclTabRead(xftab, NULL, EXIT_ON_FAIL)
         ;  streamer.tab_sym_in = tab
         ;  if ((readx_bits & MCLX_REQUIRE_CANONICAL) && !MCLV_IS_CANONICAL(tab->domain))
            mcxDie(1, caller, "need a canonical domain")
      ;  }
         mx
      =  mclxIOstreamIn
         (  xfabc
         ,     MCLXIO_STREAM_ABC
            |  MCLXIO_STREAM_MIRROR
            |  MCLXIO_STREAM_SYMMETRIC
            |  (tab ? MCLXIO_STREAM_GTAB_RESTRICT : 0)
         ,  NULL
         ,  mclpMergeMax
         ,  &streamer
         ,  EXIT_ON_FAIL
         )
      ;  tab = streamer.tab_sym_out
      ;  if (readx_bits & MCL_READX_REMOVE_LOOPS)
         mclxAdjustLoops(mx, mclxLoopCBremove, NULL)
   ;  }
      else
      {  mx = mclxReadx(xfmx, EXIT_ON_FAIL, readx_bits)
      ;  if (xftab)
         tab = mclTabRead(xftab, mx->dom_cols, EXIT_ON_FAIL)
   ;  }
      tabpp[0] = tab

   ;  if (transform)
      mclgTFexec(mx, transform)

   ;  return mx
;  }


mcxstatus mcx_dump_node
(  FILE* fp
,  const mclTab* tab
,  long idx
)
   {  unsigned n_missed = 0
   ;  if (tab)
      {  const char* label = mclTabGet(tab, idx, NULL)
      ;  if (label == tab->na->str)
            fprintf(fp, "?_%ld", idx)
         ,  n_missed = 1
      ;  else
         fputs(label, fp)
   ;  }
      else
      fprintf(fp, "%ld", idx)
   ;  return n_missed
;  }

